core/src/main/java/hudson/logging/LogRecorder.java

```diff
@@ -97,13 +99,54 @@
 public class LogRecorder extends AbstractModelObject implements Saveable {
     private volatile String name;
 
-    public final CopyOnWriteList<Target> targets = new CopyOnWriteList<>();
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("TODO")
+    /**
+     * No longer used.
+     *
+     * @deprecated use {@link #getLoggers()}
+     */
+    public final transient CopyOnWriteList<Target> targets = new CopyOnWriteList<>();
+    private List<Target> loggers = new ArrayList<>();
     private static final TargetComparator TARGET_COMPARATOR = new TargetComparator();
-    
+
+    @DataBoundConstructor
+    public LogRecorder(String name) {
+        this.name = name;
+        // register it only once when constructed, and when this object dies
+        // WeakLogHandler will remove it
+        new WeakLogHandler(handler,Logger.getLogger(""));
+    }
+
+    private Object readResolve() {
+        if (loggers == null) {
+            loggers = new ArrayList<>();
+        }
+
+        List<Target> tempLoggers = new ArrayList<>(loggers);
+
+        if (!targets.isEmpty()) {
+            loggers.addAll(targets.getView());
+        }
+        if (!tempLoggers.isEmpty() && !targets.getView().equals(tempLoggers)) {
+            targets.addAll(tempLoggers);
+        }
+        return this;
+    }
```

Comment:
My strong preference would be to add a test for this `readResolve()` method using `@LocalData` as in e.g. https://github.com/jenkinsci/email-ext-plugin/blob/master/src/test/java/hudson/plugins/emailext/ExtendedEmailPublisherDescriptorTest.java#L512-L704